### DISCUSSION TITLE
Addind/Creating  Athena Client

### DISCUSSION THREAD

**Discussion by karamokobaD** [async-aws/aws#1387] [Q&A]
Hi everyone, and think for your great and helpfull work :pray: .

Following your description/instruction for adding new Client.
I started working on Athena Client and some of his operations.
Draft PR :  [PR 1388](https://github.com/async-aws/aws/pull/1388)

I was getting an error when trying to add Athena GetQueryResults operation ( which is necessary/important ) :worried:. 
```
In PaginationGenerator.php line 312:
                                                                           
  Pagination without resultkey is not yet implemented for GetQueryResults 
```
GetQueryResults is well present in paginators-1.json and his description does not have resultkey define.
And GetQueryResultsOutput seem to be a bit more complexe than a ListShap, for his description :
 - https://awscli.amazonaws.com/v2/documentation/api/latest/reference/athena/get-query-results.html
 - https://github.com/aws/aws-sdk-php/blob/master/src/data/athena/2017-05-18/api-2.json

So, It's surely not correct way to go, but I used this quick fix/work around to generate GetQueryResultsOutput

```php
// Class src/CodeGenerator/src/Generator/OperationGenerator.php 
//ligne 29
 private const PAGINATE_STRUCTURE_SHAPE = ['GetQueryResults'];
//ligne 142 143
if (null !== $operation->getOutput()) {  
 $resultClass = $this->resultGenerator->generate($operation);  
 $isPaginateListShape = false === \in_array($operation->getName(), self::PAGINATE_STRUCTURE_SHAPE);  //quick fix
if (null !== $operation->getPagination() && $isPaginateStructure) {  
 $this->paginationGenerator->generate($operation);  
}  
  
 $method->setReturnType($resultClass->getFqdn());  
$classBuilder->addUse($resultClass->getFqdn());  
}
```

And I also have/got an deprecated Class : `CalculationConfiguration` but could not find information about new Class to used instead yet. 

Does anyone have a better idea on how to properly fix / work around that ?

**Comment by jderusse** [comment]
It looks like it's a mistake in the AWS declaration. 
Are you able to give a try to the official AWS SDK implementation and try to iterate over the results.

Result list is hidden in the `GetQueryResult.Resultset.Rows` properties, but nothing references this in the .json files.

I suggest patching the pagination.json file: AsyncAws code generator already provides the glue for this. You can add a `patch` section in the manifest file.

```

            "patches": {
                "pagination": [
                    {
                        "value": "ResultSet.Rows",
                        "op": "add",
                        "path": "/pagination/GetQueryResults/result_key"
                    }
                ]
            }
```

**Comment by karamokobaD** [comment]
Hi :wave: 

I tried the patch you suggest, but I got the error  ` Pagination without resultkey is not yet implemented for GetQueryResults `
Also the official AWS SDK implement Result return a  ResultSetMetadata with a column structure information with can also be usefull.
It's more structured like this , if I am not mistaken
```
GetQueryResultsOutput 
        →  UpdateCount
        →  ResultSet
            – Rows [] 
              + Data []
            – ResultSetMetadata
              + ColumnInfo []
        →  NextToken
```
 
Are you able to give a try to the official AWS SDK implementation and try to iterate over the results. ==> The example code I add (athena.md)  about how submit a query and retrieves the result show a way to do it.

**Reply by jderusse** [reply]
> Pagination without resultkey is not yet implemented for GetQueryResults

Because the generator caches the json files, you need to clear cache (ie. `rm -rf .cache`)

> AWS SDK implement Result return a ResultSetMetadata

Yes, so does AsyncAws. Pagination is all about iterating over the result. It's kind of a second way to consume the data:

You can use properties

```php
$result  = $client->getQueryResult();
dump($result->getUpdateCount());
dump($result->getResultSet()->getRows());
```

or you can iterate over the result (this is 

```php
$result  = $client->getQueryResult();
foreach ($result as $row) {
}
```

behind the wood, this is equivalent to
```php
$result  = $client->getQueryResult();
foreach ($result->getResultSet()->getRows() as $row) {
  // + magic to fetch the next page when retrieving the last item of the page
}
```

**Comment by karamokobaD** [comment]
Thanks for this explanation. 
Like you said clear the .cache and the GetQueryResults was re-generate the implementation of IteratorAggregate  :muscle:

**Comment by karamokobaD** [comment]
Hi @jderusse :wave:

I have one CI check remained : 

 - [Psalm](https://github.com/async-aws/aws/actions/runs/4698074438/jobs/8329912551?pr=1388#logs) due to the GetQueryResults ouput generation with the "patchs". 
    
But one of the contirib rules is that : " code cannot be manually changed. ". 
And even if we changed it manually we will get an error with the generated ouput not change check. 

For now  we just add it to paslm.baseline.xml file

**Comment by karamokobaD** [comment]
@jderusse thank a lot for the time you took to review,  validate and merge our PR :pray: :pray:  

I also want to know if we can already require Athena client (composer require ...) like others client ?

**Reply by jderusse** [reply]
Hello, I just releqase the first version of the `async-aws/athena`.
You can run `composer req async-aws/athena`

**Reply by karamokobaD** [reply]
:fireworks:   thank a lot
